README.md
# SafeIntId

Uncoordinated unique ID that fits [MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) on web platform for long enough time.

## Existing solutions

There are [a lot](https://medium.com/geekculture/the-wild-world-of-unique-identifiers-uuid-ulid-etc-17cfb2a38fce)
of well-designed unique IDs. Few well-known ones are, alphabetically:  
[Cuid2](https://github.com/paralleldrive/cuid2),
[KSUID](https://github.com/segmentio/ksuid),
[Nano ID](https://github.com/ai/nanoid),
[ObjectId](https://www.mongodb.com/docs/manual/reference/method/ObjectId/),
[Push ID](https://firebase.blog/posts/2015/02/the-2120-ways-to-ensure-unique_68#whats-in-a-push-id),
[Sharding ID](https://instagram-engineering.com/sharding-ids-at-instagram-1cf5a71e5a5c),
[Snowflake ID](https://en.wikipedia.org/wiki/Snowflake_ID),
[Sonyflake](https://github.com/sony/sonyflake),
[ULID](https://github.com/ulid/spec),
[UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier),
[XID](https://github.com/rs/xid).

These IDs are great for their use cases, but none meets our special requirements.

## Requirements

* ID should fit signed 64-bit int as in [Isar Database ID](https://isar.dev/schema.html#isar-id).
* ID should fit +/- (2⁵³-1) as in [safe integers on web platform](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger#description).
* IDs should be generated at multiple places without coordination over the network
  as in offline-first app running on multiple devices.
* IDs created at different moments of time should have zero probability of collision,
  not just small one.
* IDs should be unique as much as possible given the constraints above.

## Nice to have

* IDs could be sorted by their created-at timestamp.
* This timestamp could be extractable from ID.
* ID could be URL-friendly.
    * Max ID is `9007199254740991` which is URL-friendly as is.
    * If you need even shorter string, you can encode and decode ID, e.g:
        * `9007199254740991.toRadixString(36) == "2gosa7pa2gv"`
        * `int.parse("2gosa7pa2gv", radix: 36) == 9007199254740991`
* ID could optionally contain an auto-incrementing counter instead of a random part.

## Not required

* We don't require ID to be highly unpredictable:
  we'd use auto-increment ID, but it would collide without coordination.

## Design and usage

* Import:
    ```dart
    import 'package:safe_int_id/safe_int_id.dart';
    ```
* Get a new ID:
    ```dart
    final id = safeIntId.getId();
    ```
    * ID is a positive integer less or equal than 2⁵³-1 for some number of years.
    * We could have 54 bits by using negative integers too,
      but this would increase complexity and reduce compatibility.
    * So we have 53 bits.
* ID contains, by default:
    * 43 bits of timestamp:
        * Milliseconds since the UTC start of the first year you use this ID,
          2023 by default, configurable:
            ```dart
            final customId = SafeIntId(firstYear: 2024);
            ```
        * To get `DateTime` when given `id` was created at:
            ```dart
            safeIntId.getCreatedAt(id) is DateTime
            ```
        * 43 bits give us more than 278 years of millisecond-precise timestamps:
            ```dart
            pow(2, 43) / (1000*60*60*24*365.2425) > 278
            ```
        * So for the default first year 2023 the last safe year is 2300, good enough.
    * 10 bits of random:
        * Non-secure random generator is used by default:
          it is always available and is faster than the secure one.
        * If you prefer secure random generator, it is configurable:
            ```dart
            final customId = SafeIntId(random: Random.secure());
            ```
        * 10 random bits give 1024 possible values per millisecond, configurable
          in exchange for how many years since `firstYear` this ID will be safe:
            ```dart
            final customId = SafeIntId(randomValues: 2048);
            assert(customId.safeYears == 139);
            assert(customId.lastSafeYear == 2161);
            ```

```
| random |                  default firstYear 2023 |
|   bits | randomValues | safeYears | lastSafeYear |
| ------ | ------------ | --------- | ------------ |
|      8 |          256 |      1114 |         3136 |
|      9 |          512 |       557 |         2579 |
|     10 | default 1024 |       278 |         2300 |
|     11 |         2048 |       139 |         2161 |
|     12 |         4096 |        69 |         2091 |
|     13 |         8192 |        34 |         2056 |
|     14 |        16384 |        17 |         2039 |
|     15 |        32768 |         8 |         2030 |
|     16 |        65536 |         4 |         2026 |
```

* Probability of generating more than one ID with the same value
  is [approximately N²/2M](https://en.wikipedia.org/wiki/Birthday_problem#Square_approximation),
  where:
    * M is `randomValues` - how many possible values we have.
    * N is how many IDs per millisecond we generate.
    * For N = 1 or less the probability is exactly zero for any M.
    * For N = 2 and default M = 1024 we have probability 0.002 = 0.2%.
    * 2 IDs per millisecond means 2K IDs per second,
      120K IDs per minute, 7.2M IDs per hour.
    * To get 1% probability of collision with default M = 1024 we need to generate
      `sqrt(2*1024*0.01)` = 5 IDs per millisecond,
      5K IDs per second, 272K IDs per minute, 16M IDs per hour.
    * Please check the table below to adjust `SafeIntId(randomValues: M)` if needed.
    * This approximation works well for probabilities of 50% or less,
      so the higher ones are not shown.

```
| M     | N                                                                                |
|       | 1 | 2      | 5     | 10   | 20   | 30  | 40  | 60  | 80  | 100 | 150 | 200 | 250 |
|------ | - | ------ | ----- | ---- | ---- | --- | --- | --- | --- | --- | --- | --- | --- |
|   256 | 0 | 0.8%   | 4.9%  |  20% |      |     |     |     |     |     |     |     |     |
|   512 | 0 | 0.4%   | 2.4%  |  10% |  39% |     |     |     |     |     |     |     |     |
|  1024 | 0 | 0.2%   | 1.2%  | 4.9% |  20% | 44% |     |     |     |     |     |     |     |
|  2048 | 0 | 0.1%   | 0.6%  | 2.4% |  10% | 22% | 39% |     |     |     |     |     |     |
|  4096 | 0 | 0.05%  | 0.3%  | 1.2% |   5% | 11% | 20% | 44% |     |     |     |     |     |
|  8192 | 0 | 0.02%  | 0.15% | 0.6% |   2% |  5% | 10% | 22% | 39% |     |     |     |     |
| 16384 | 0 | 0.01%  | 0.08% | 0.3% |   1% |  3% |  5% | 11% | 20% | 31% |     |     |     |
| 32768 | 0 | 0.01%  | 0.04% | 0.2% |   1% |  1% |  2% |  5% | 10% | 15% | 34% |     |     |
| 65536 | 0 | 0.003% | 0.01% | 0.1% | 0.3% |  1% |  1% |  3% |  5% |  8% | 17% | 31% | 48% |
```

* If you generate a lot of IDs on the same device,
  use `incId()` instead of `getId()`:
    * `incId()` increments a counter instead of using a random value.
    * This counter resets to zero each millisecond, and blocks reaching `randomValues` (1024 by default).
    * If a hot loop for less than 1 millisecond burns,
      use `await incIdAsync()` or increase `randomValues` in exchange for `safeYears`.
    * This way you can get an increasing sequence of unique IDs.


CHANGELOG.md
# 1.1.1

* Added test for `incIdAsync`.
* Fixed docs re "increasing sequence".

# 1.1.0

* Added `incId()` incrementing counter instead of using random value.
  Counter resets to zero each millisecond, blocks reaching `randomValues`.
* Added `incIdAsync()` if hot loop for less than 1 millisecond burns.

# 1.0.0

* Initial stable release.


test/safe_int_id_test.dart
import "dart:io";
import "package:safe_int_id/safe_int_id.dart";
import "package:test/test.dart";

void main() {
  group("By default", () {
    test("ID and createdAt are generated as expected", () {
      final start = DateTime.now();
      final id = safeIntId.getId();
      final createdAt = safeIntId.getCreatedAt(id);
      final diffMillis = createdAt.difference(start).inMilliseconds;
      expect(diffMillis, inInclusiveRange(0, 10));
    });

    test("IDs don't collide when generating 1 or less IDs per millisecond", () {
      final previous = <int>{};
      for (int i = 0; i < 1000; i++) {
        final id = safeIntId.getId();
        expect(previous, isNot(contains(id)));
        previous.add(id);
        expect(previous, contains(id));
        sleep(Duration(milliseconds: 1));
      }
    });

    test("IDs are increasing when using `incId`", () {
      int previous = 0;
      for (int i = 0; i < 10000; i++) {
        final id = safeIntId.incId();
        expect(previous, lessThan(id));
        previous = id;
      }
    });

    test("IDs are increasing when using `await incIdAsync`", () async {
      int previous = 0;
      for (int i = 0; i < 10000; i++) {
        final id = await safeIntId.incIdAsync();
        expect(previous, lessThan(id));
        previous = id;
      }
    });

    test("config is set as expected", () {
      expect(safeIntId.firstYear, equals(2023));
      expect(safeIntId.randomValues, equals(1024));
      expect(safeIntId.safeYears, equals(278));
      expect(safeIntId.lastSafeYear, equals(2300));
    });
  });

  group("With custom config", () {
    test("bigger [firstYear] increases [lastSafeYear]", () {
      final customId = SafeIntId(firstYear: 2024);
      expect(customId.safeYears, equals(278));
      expect(customId.lastSafeYear, equals(2301));
    });

    test("bigger [randomValues] decreases [safeYears] and [lastSafeYear]", () {
      final customId = SafeIntId(randomValues: 2048);
      expect(customId.safeYears, equals(139));
      expect(customId.lastSafeYear, equals(2161));
    });

    test("smaller [randomValues] increases [safeYears] and [lastSafeYear]`",
        () {
      final customId = SafeIntId(randomValues: 512);
      expect(customId.safeYears, equals(557));
      expect(customId.lastSafeYear, equals(2579));
    });

    test(
        "zero [randomValues] increases [safeYears] and [lastSafeYear] a lot, "
        "while [getId] and [getCreatedAt] are still working", () {
      final customId = SafeIntId(randomValues: 0);
      expect(customId.safeYears, equals(285426));
      expect(customId.lastSafeYear, equals(287448));

      final id = customId.getId();
      expect(id, greaterThan(0));

      final diffMillis = customId
          .getCreatedAt(id)
          .difference(DateTime(customId.firstYear))
          .inMilliseconds;
      expect(diffMillis, greaterThan(0));
    });
  });
}


example/safe_int_id_example.dart
/// Example of how to use `SafeIntId`.
/// Run it with `dart run --enable-asserts example/safe_int_id_example.dart`

import "dart:math";
import "package:safe_int_id/safe_int_id.dart";

void main() {
  // Get a new uncoordinated unique ID
  // that fits `Number.MAX_SAFE_INTEGER` on web platform for long enough time.
  final id = safeIntId.getId();
  print("id: $id");

  // Get [DateTime] when given [id] was created at.
  final createdAt = safeIntId.getCreatedAt(id, isUtc: true);
  print("createdAt: $createdAt");

  // Increment a counter instead of using a random value.
  // Counter resets to zero each millisecond, blocks reaching max (1024 by default).
  final id1 = safeIntId.incId();
  final id2 = safeIntId.incId();
  assert(id1 < id2);

  // Custom configuration, defaults are:
  final customId = SafeIntId(
    firstYear: 2023, // The first year you use this ID.
    random: Random(), // Random generator, e.g. `Random.secure()`
    randomValues: 1024, // Number of possible random values per millisecond.
    // Makes ID more unique in exchange for [safeYears].
  );

  // How many years since [firstYear] this ID will be safe:
  assert(customId.safeYears == 278);

  // Last year this ID will be safe:
  assert(customId.lastSafeYear == 2300);

  // Custom configuration has the same [getId] and [getCreatedAt] methods:
  assert(!customId.getCreatedAt(customId.getId()).isBefore(createdAt));
}


lib/safe_int_id.dart
/// Uncoordinated unique ID
/// that fits `MAX_SAFE_INTEGER` on web platform for long enough time.
library safe_int_id;

import "dart:math";

/// `SafeIntId` config and methods. Please see `README.md` for details.
class SafeIntId {
  /// The first year you use this ID.
  late final int firstYear;

  /// Milliseconds from Epoch to the UTC start of the [firstYear].
  late final int _firstYearMillis;

  /// Random generator.
  late final Random random;

  /// Number of possible random values per millisecond.
  /// Makes ID more unique in exchange for [safeYears].
  late final int randomValues;

  /// How many years since [firstYear] this ID will be safe.
  late final int safeYears;

  /// Last year this ID will be safe.
  late final int lastSafeYear;

  /// Counter used by [incId].
  int _counter = 0;

  /// The previous number of milliseconds in [incId].
  int _previousMillis = 0;

  /// Create new `SafeIntId` generator.
  SafeIntId({
    this.firstYear = 2023,
    Random? random,
    int randomValues = 1024,
  }) {
    _firstYearMillis = DateTime.utc(firstYear).millisecondsSinceEpoch;
    this.random = random ?? Random();
    this.randomValues = max(randomValues, 1);
    safeYears =
        pow(2, 53) ~/ (this.randomValues * 1000 * 60 * 60 * 24 * 365.2425);
    lastSafeYear = firstYear + safeYears - 1;
  }

  /// Get a new uncoordinated unique ID
  /// that fits `Number.MAX_SAFE_INTEGER` on web platform for long enough time.
  int getId() {
    final millis = DateTime.now().millisecondsSinceEpoch - _firstYearMillis;
    return millis * randomValues + random.nextInt(randomValues);
  }

  /// The same as [getId] but increments counter instead of using random value.
  /// Counter resets to zero each millisecond, blocks reaching [randomValues].
  /// Use `await incIdAsync()` if hot loop for less than 1 millisecond burns.
  int incId() {
    while (true) {
      final millis = DateTime.now().millisecondsSinceEpoch - _firstYearMillis;
      if (_previousMillis != millis) {
        _counter = 0;
        _previousMillis = millis;
      }
      if (_counter >= randomValues) continue; // `sleep` is N/A on web platform.
      return millis * randomValues + _counter++;
    }
  }

  /// The same as [incId] but async.
  Future<int> incIdAsync() async {
    while (true) {
      final millis = DateTime.now().millisecondsSinceEpoch - _firstYearMillis;
      if (_previousMillis != millis) {
        _counter = 0;
        _previousMillis = millis;
      }
      if (_counter >= randomValues) {
        await Future.delayed(Duration(microseconds: 100));
        continue;
      }
      return millis * randomValues + _counter++;
    }
  }

  /// Get [DateTime] when given [id] was created at.
  DateTime getCreatedAt(int id, {bool isUtc = false}) =>
      DateTime.fromMillisecondsSinceEpoch(id ~/ randomValues + _firstYearMillis,
          isUtc: isUtc);
}

/// `SafeIntId` with default config.
final safeIntId = SafeIntId();
